lib/components/canvas/_asset_cache.dart

```diff
@@ -83,6 +83,18 @@ class OrderedAssetCache {
 
   final List<Object> _cache = [];
 
+  /// Maps a file path to already added assets.
+  final Map<String, int> _fileIndices = {};
+
+  void addAsset(File? file, int index) {
+    if (file == null) return;
+    _fileIndices[file.path] = index;
```

Comment:
use _filePdfIndices

Reply:
May be addPdfAssetIndice  will be even better, because we store only indice not asset itself.